src/main/java/de/tum/in/www1/artemis/repository/hestia/CoverageReportRepository.java

```diff
@@ -0,0 +1,106 @@
+package de.tum.in.www1.artemis.repository.hestia;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import org.springframework.stereotype.Repository;
+
+import de.tum.in.www1.artemis.domain.hestia.CoverageFileReport;
+import de.tum.in.www1.artemis.domain.hestia.CoverageReport;
+import de.tum.in.www1.artemis.domain.hestia.TestwiseCoverageReportEntry;
+import de.tum.in.www1.artemis.service.connectors.bamboo.dto.TestwiseCoverageReportDTO;
+import de.tum.in.www1.artemis.web.rest.errors.EntityNotFoundException;
+
+/**
+ * Spring Data JPA repository for the CoverageReport entity.
+ */
+@Repository
+public interface CoverageReportRepository extends JpaRepository<CoverageReport, Long> {
+
+    default CoverageReport findByIdElseThrow(Long coverageReportId) {
+        var optionalReport = findById(coverageReportId);
+        return optionalReport.orElseThrow(() -> new EntityNotFoundException("Coverage Report", coverageReportId));
+    }
+
+    Optional<CoverageReport> findCoverageReportBySubmissionId(@Param("submissionId") Long submissionId);
+
+    default CoverageReport findCoverageReportBySubmissionIdElseThrow(Long submissionId) {
+        var optionalReport = findCoverageReportBySubmissionId(submissionId);
+        return optionalReport.orElseThrow(() -> new EntityNotFoundException("Coverage Report", submissionId));
+    }
+
+    @Query("""
+            SELECT DISTINCT r FROM CoverageReport r
+            LEFT JOIN FETCH r.fileReports f
+            LEFT JOIN FETCH f.testwiseCoverageEntries
+            WHERE r.id = :#{#coverageReportId}
+            """)
+    Optional<CoverageReport> findCoverageReportByIdWithEagerFileReportsAndEntries(@Param("coverageReportId") Long coverageReportId);
+
+    default CoverageReport findCoverageReportByIdWithEagerFileReportsAndEntriesElseThrow(Long coverageReportId) {
+        var optionalReport = findCoverageReportByIdWithEagerFileReportsAndEntries(coverageReportId);
+        return optionalReport.orElseThrow(() -> new EntityNotFoundException("Coverage Report", coverageReportId));
+    }
+
+    /**
+     * Transforms the testwise coverage report dtos to CoverageFileReports (without the test case attribute) mapped by the test case name.
+     * This method maps the file reports to primitive test case names because the test case is not present in the database
+     * on creating the entities from the dtos.
+     * @param coverageReports the coverage report dtos
+     * @return coverage file reports mapped by the test case name
+     */
+    default Map<String, Set<CoverageFileReport>> createTestwiseCoverageFileReportsWithoutTestsByTestCaseName(List<TestwiseCoverageReportDTO> coverageReports) {
```

Comment:
Is there any reason why this is a method in the repository and not a service?
The repository should only be used for database-related operations. In fact, I don't see a single database operation in this method.